### PR TITLE
Generative kaleidescope

### PR DESCRIPTION
**Pull request by kfahn22**
I have adapted this [shader](https://www.shadertoy.com/view/ctByWz) to create a function that generates a kaleidoscope  pattern.  I think it creates some pretty cool effects, but you can decide.  https://editor.p5js.org/kfahn/sketches/I29DSDQOS
kaleidescope

### CODE REVIEW

File: generative/kaleidescope.hlsl [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef FNC_KALEIDESCOPE
+#define FNC_KALEIDESCOPE
+float kaleidescope( float2 st, float speed, float zoom, float m, float n, int N) {
+    #ifdef CENTER_2D
+    st -= CENTER_2D;
+    #else
+    st -= 0.5;
+    #endif
+    vec3 r = u_resolution.xxx;
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.hlsl:25]: `vec3` > `float3`

File: generative/kaleidescope.glsl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef FNC_KALEIDESCOPE
+#define FNC_KALEIDESCOPE
+float kaleidescope( vec2 st, float speed, float zoom, float m, float n, int N) {
+    #ifdef CENTER_2D
+    st -= CENTER_2D;
+    #else
+    st -= 0.5;
+    #endif
+    vec3 r = u_resolution.xxx;
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.glsl:25]: we can not warranty the presence of `u_resolution` uniform

File: generative/kaleidescope.glsl [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    vec3 uv = vec3((2.0 * st.xyy- r)/r.x * zoom);
+    uv.z = sineIn(u_time*speed);
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.glsl:27]: we can not warranty the presence of `u_time` uniform

Comment (patriciogonzalezvivo) [inline generative/kaleidescope.glsl:27]: What's `sineIn` ? is not native, right?

File: generative/kaleidescope.hlsl [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    vec3 uv = vec3((2.0 * st.xyy- r)/r.x * zoom);
+    uv.z = sineIn(u_time*speed);
+    uv *= 0.35;
+    #ifdef PLATFORM_WEBGL
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.hlsl:29]: No need to do WEBGL version in HLSL

File: generative/kaleidescope.hlsl [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    for (int i = 0; i< 10; i++) {
+        if (i >= N) break;
+    #else
+    for (int i = 0; i< N; i++) {
+    #endif
+		uv = abs( uv ) / lengthSq( uv ) - m; 
+        uv = length( uv )*( uv + n);
+	}
+    uv = float3(lengthSq( uv ) * 0.5) ;
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.hlsl:38]: `float3()` takes 3 arguments

File: generative/kaleidescope.hlsl [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+#ifndef FNC_KALEIDESCOPE
+#define FNC_KALEIDESCOPE
+float kaleidescope( float2 st, float speed, float zoom, float m, float n, int N) {
+    #ifdef CENTER_2D
+    st -= CENTER_2D;
+    #else
+    st -= 0.5;
+    #endif
+    vec3 r = u_resolution.xxx;
+    vec3 uv = vec3((2.0 * st.xyy- r)/r.x * zoom);
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.hlsl:26]: `vec3` > `float3`
`float3` takes 3 arguments

File: generative/kaleidescope.hlsl [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#include "../math/lengthSq.hlsl"
+
+/*
+description: |
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.hlsl:4]: Add `original_author` field

File: generative/kaleidescope.hlsl [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+#define FNC_KALEIDESCOPE
+float kaleidescope( float2 st, float speed, float zoom, float m, float n, int N) {
+    #ifdef CENTER_2D
+    st -= CENTER_2D;
+    #else
+    st -= 0.5;
+    #endif
+    vec3 r = u_resolution.xxx;
+    vec3 uv = vec3((2.0 * st.xyy- r)/r.x * zoom);
+    uv.z = sineIn(u_time*speed);
```
Comment (patriciogonzalezvivo) [inline generative/kaleidescope.hlsl:27]: What's `sineIn()` ?

**Review by patriciogonzalezvivo — CHANGES_REQUESTED** [review]
this feels more like a `lygia/space` operation that a `lygia/generative`
Consider returning a `vec2` with a mirrored/kaleidoscopic space

**Comment by kfahn22** [comment]
Thanks for the feedback!  I have tried to address all of your comments. The files have been moved to space and the function now returns a vec2/float2.  Please let me know if you have any more feedback/requested changes.

File: space/kaleidescope.glsl [hunk 9]
```diff
@@ -0,0 +1,43 @@
+#include "..animation/easing/sineIn.glsl"
```
Comment (patriciogonzalezvivo) [inline space/kaleidescope.glsl:1]: correct path is `#include "../animation/easing/sine.glsl"`

File: space/kaleidescope.glsl [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    * For a more classic kaleidescope effect, use uv.x, m = 1.75 and n = 1.15;
+    * For a "tie-dye" look, use, uv.x, m = 1.5 and n = 1.3
+    * Calling the kaleidescope twice with different parameters creates some interesting effects
+    * Clamping the returned value is useful if colors are to bright
+use: kaleidescope(<vec2> st, <vec2> pixel, <float> t, <float> zoom, <float> m, <float> n, <int> interations)
+*/
+
+#ifndef FNC_KALEIDESCOPE
+#define FNC_KALEIDESCOPE
+vec2 kaleidescope( vec2 st, vec2 pixel, float t, float zoom, float m, float n, int N) {
```
Comment (patriciogonzalezvivo) [inline space/kaleidescope.glsl:21]: I believe maybe the right name is "kaleidoscope" ?

**Review by patriciogonzalezvivo — CHANGES_REQUESTED** [review]
This doesn't seams to return a `vec2` coordinate space. Seams to specific of an algorithm.
Could be generalize so it mirrors and fract an 2D space into N sections like kaleidoscopes do?

**Comment by kfahn22** [comment]
I am not sure if this is what you have in mind?  Or more subdivisions?
https://editor.p5js.org/kfahn/sketches/3ZmPyZVjv
Screenshot 2023-08-23 at 4 12 09 PM

**Pull request closed without merging**